### ISSUE TITLE
chapter01/1.3 - URLify/urlify.js

### ISSUE THREAD

**Issue by jarkin13** [careercup/CtCI-6th-Edition-JavaScript#20]
For this problem, why wouldn't you do something along the lines of:

`var uglify = function(input) {
  input = input.trim();
  var url = input.split(' ').join('%20');
  return url;
}`

I know it says that it provides with the true length of the string, but would something a long these lines not be acceptable?

**Comment by skawaguchi** [comment]
You could also use Regex. But the exercise could also be to see how you would build this out in a more general way. I assumed that this was simple for the reason to see how would would modularize your code and go with a faster runtime, rather than to focus on how you would manipulate strings in a language that didn’t have these tools.
```js
const charMap = {
  ‘ ‘: ‘%20’
};

function replaceCharacter(str, targetChar, replacementChar) {
  const charRegex = new RegExp(targetChar, ‘gi’);

  return str.replace(charRegex, replacementChar);
}

function urlify(url) {
  let modifiedURL = url.trim();

  for(let charToReplace in charMap) {
    const replacementChar = charMap[charToReplace];

    modifiedURL = replaceCharacter(modifiedURL, charToReplace, replacementChar);
  }
  return modifiedURL;
}
```

Of course, if we are pretending there are no language features that trivialize the problem or we need to account for space (maybe the target device is embedded with insanely small resources), then we could do something along the lines of the existing code or the solution from the book.

Basically it seems your interviewer can make it as hard as they want to, so we should probably consider all cases.

**Comment by Dg-0987** [comment]
Apparently there is an issue with this solution because when you run it the result it returns is missing some characters: `Mr%20hohn%20ith%20`.
Also without using any JS specific methods/properties you could still simplify this even more.

**Comment by irontitan76** [comment]
Since I don't like recreating the wheel: 

`function urlify(str, length) {
var arr = str.trim().split('');

arr.forEach((item, i) => {
if ( item === ' ' ) arr[i] = '%20';
});

return arr.join('');
};`

**Comment by Elaniobro** [comment]
I came here to write, you could simply use regex, I do like your solution @jarkin13, I did not think to use `trim();`

`const URLify = (string) => string.replace(/[ \t]+$/g, '').replace(/[ ]/g, '%20');`

**Comment by AmitJoki** [comment]
I think the following is a general solution and that which is more elegant.

```
var urlify = function(str, length) {
  var strArr = [];
  for(var i = 0; i < length; i++) {
     strArr[i] = str[i] == ' ' ? '%20' : str[i]; 
  }
  return strArr.join('');
};
```

I've submitted a PR for the same.
